multiaccounts/settings_wallet/database.go

```diff
@@ -0,0 +1,153 @@
+	if tx == nil {
+		return nil
+	}
+	_, err := tx.Exec("UPDATE settings SET wallet_token_preferences_group_by_community = ? WHERE synthetic_id = 'id'", value)
+	return err
+}
+
+func (db *WalletSettings) GetTokenGroupByCommunity() (result bool, err error) {
+	query := "SELECT wallet_token_preferences_group_by_community FROM settings WHERE synthetic_id = 'id'"
+	err = db.db.QueryRow(query).Scan(&result)
+	if err != nil {
+		return false, err
+	}
+	return result, err
+}
+
+func (db *WalletSettings) UpdateTokenPreferences(preferences []TokenPreferences, groupByCommunity bool, testNetworksEnabled bool, clock uint64) error {
+	if len(preferences) == 0 {
+		return errors.New("tokens: trying to create custom order with empty list")
+	}
+
+	tx, err := db.db.Begin()
+	if err != nil {
+		return err
+	}
+
+	var mainError error = nil
+
+	defer func() {
+		if mainError == nil {
+			err = tx.Commit()
+			return
+		}
+		_ = tx.Rollback()
+	}()
+
+	_, err = tx.Exec("DELETE FROM token_preferences WHERE testnet = ?", testNetworksEnabled)
+	if err != nil {
+		mainError = err
+		return err
+	}
+
+	for _, p := range preferences {
+		if p.Position < 0 {
+			mainError = errors.New("tokens: trying to create custom order with negative position")
+			return mainError
+		}
+		_, err := tx.Exec("INSERT INTO token_preferences (key, position, group_position, visible, community_id, testnet) VALUES (?, ?, ?, ?, ?, ?)", p.Key, p.Position, p.GroupPosition, p.Visible, p.CommunityID, testNetworksEnabled)
+		if err != nil {
+			mainError = err
+			return err
+		}
+	}
+
+	if groupByCommunity {
+		// Find community tokens without group position
+		// Group position can be -1 if it wasn't created yet. Values must be consitstent across all tokens
+		rows, err := tx.Query(`SELECT COUNT(*) FROM token_preferences WHERE testnet = ? AND group_position = -1 AND community_id != '' AND visible GROUP BY community_id HAVING COUNT(*) > 0`, testNetworksEnabled)
+		if err != nil {
+			mainError = err
+			return err
+		}
+		if rows.Next() {
+			mainError = errors.New("tokens: not all community tokens have assigned the group position")
+			return mainError
+		}
+	}
+
+	err = db.setTokenGroupByCommunity(tx, groupByCommunity)
+	if err != nil {
+		mainError = err
+		return err
+	}
+
+	err = db.setClockOfLastTokenPreferencesChange(tx, clock)
+	if err != nil {
+		mainError = err
+		return err
+	}
+	return nil
+}
+
+func (db *WalletSettings) GetTokenPreferences(testNetworksEnabled bool) ([]TokenPreferences, error) {
+	rows, err := db.db.Query("SELECT key, position, group_position, visible, community_id FROM token_preferences WHERE testnet = ?", testNetworksEnabled)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	var result []TokenPreferences
+
+	for rows.Next() {
+		token := TokenPreferences{}
+		err := rows.Scan(&token.Key, &token.Position, &token.GroupPosition, &token.Visible, &token.CommunityID)
+		if err != nil {
+			return nil, err
+		}
+
+		result = append(result, token)
+	}
+
```

Comment:
@Cuteivist please add this check:
```suggestion
	if err := rows.Err(); err != nil {
		return nil, err
	}
```

Reply:
Thanks! Fixed.